### ISSUE TITLE
twisted.web has no protection against HTTP response-splitting attacks

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3770]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| ivank reported|
|-|-|
|Trac ID|trac#3770|
|Type|defect|
|Created|2009-04-11 16:27:22Z|

twisted.web should check for \r or \n in header values before writing out the headers. This can prevent HTTP response-splitting attacks.

An application developer should be able to assume that setting one header with possibly malicious data doesn't have consequences beyond that one header.

I believe the right place to do this is in twisted.web.http.Request.write, because that method assumes that \r\n will be separating the headers.

Attachments:

* [no-response-splitting-3770.1.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f9b/f9b07ad2b9436c277a08bf5abb260bf44512cb9c/2d3a0fa6966c09e53be1a92187462a3754e45975.diff) (8360 bytes) - added by ivank on 2009-04-13 06:17:40Z - return Internal Server Error instead of writing out strings with \r or \n
* [3770-01-validation-in-Headers-and-server-logging.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f9b/f9b07ad2b9436c277a08bf5abb260bf44512cb9c/f3ccea4d384883740b6c8a5031cfe276124c0e84.patch) (5245 bytes) - added by ivank on 2011-04-18 10:59:42Z - 
* [3770-02-validation-in-Headers-and-server-logging.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f9b/f9b07ad2b9436c277a08bf5abb260bf44512cb9c/84b3ba6943426fb25bfe311ee2dd36262cba8e85.patch) (5351 bytes) - added by ivank on 2011-04-18 11:04:43Z - forgot to loseConnection after the _sendError

<details><summary>Searchable metadata</summary>

```
trac-id__3770 3770
type__defect defect
reporter__ivank ivank
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__duplicate duplicate
component__web web
keywords__security security
time__1239467242000000 1239467242000000
changetime__1552932706056812 1552932706056812
version__None None
owner__ivank ivank
cc__jknight cc__ivank
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|ivank set owner to ivank|
|-|-|

I don't think the request should succeed at all if any response headers, cookies, or code_message contain \r or \n. But I don't know if this is really correct, or if it is, which HTTP response code to return.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

If an application tries to set an invalid value for some field, it would be good if an exception were thrown at that application code.  If this ends up getting back to the web server protocol code somehow, then the existing error handling code should take care to log it and return a 500 to the client.

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|ivank commented|
|-|-|

This is not ready for review. I'm going to run it for a while, and clean everything up soon.

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|ivank commented|
|-|-|

My "simple" solution in the patch is incorrect. I found this in RFC 2616:

"Header fields can be extended over multiple lines by preceding each extra line with at least one SP or HT."

I might try again with something that allows this.

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|ivank commented|
|-|-|

Based on how `http_headers.Headers` is used in `t.web.http.Request.write` and `t.web._newclient.Request._writeHeaders`, I decided to move HTTP header value validation into `Headers`.  I changed `Headers` to allow newlines only when followed by a space or tab character.

Note that while header value validation is performed early (as exarkun suggested), I also had to change `t.web.http.Request.write` to return a 500 and log a message if the `etag` or `cookies` are invalid.

This patch breaks a test that calls `setRawHeaders("foo", [None])` (see below).  The comment in the test suggests this must be supported for compatibility, but I'm not sure whether this must be supported in `Headers`, or just in `_DictHeaders`.  I don't know how to proceed.  exarkun's help would be appreciated.

I tested most of the code I added manually, but I'll add tests if the changes sound reasonable to a reviewer.

Another thing: `t.web.test.test_http.DummyHTTPHandler` sets a header value to an int.  Must we support setting non-`str` header values?  The `Headers` docstrings say only `str`s are allowed.

The traceback from the failing test:

```
[ERROR]
Traceback (most recent call last):
  File "/opt/Python-x86/lib/python2.7/site-packages/twisted/web/test/test_http_headers.py", line 577, in test_setdefaultDefault
    self.assertIdentical(wrapper.setdefault("foo"), None)
  File "/opt/Python-x86/lib/python2.7/site-packages/twisted/web/http_headers.py", line 91, in setdefault
    return DictMixin.setdefault(self, name, value)
  File "/opt/Python-x86/lib/python2.7/UserDict.py", line 130, in setdefault
    self[key] = default
  File "/opt/Python-x86/lib/python2.7/site-packages/twisted/web/http_headers.py", line 53, in __setitem__
    self._headers.setRawHeaders(key, [value])
  File "/opt/Python-x86/lib/python2.7/site-packages/twisted/web/http_headers.py", line 255, in setRawHeaders
    self._checkHeaderValue(value)
  File "/opt/Python-x86/lib/python2.7/site-packages/twisted/web/http_headers.py", line 227, in _checkHeaderValue
    "instance of %r instead" % (value, type(value)))
exceptions.TypeError: Header value None should be a str but found instance of <type 'NoneType'> instead

twisted.web.test.test_http_headers.HeaderDictTests.test_setdefaultDefault
```

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|ivank set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|ivank commented|
|-|-|

Also, instead of logging and sending 500 in `.write`, I could throw an exception if `etag` or `cookies` is invalid, by checking them early in `.write`.  Would that be preferable?  What kind of exception?  Two new ones?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set owner to ivank|
|-|-|

What if multiline values got escaped by Headers (insert the space or tab as necessary) instead of outright rejected?  It seems like the caller shouldn't have to know about things like _Header fields can be extended over multiple lines by preceding each extra line with at least one SP or HT_ - that's why we have a protocol implementation.  That would half answer the question about etag and cookies, but I guess there's still a problem if a non-str is supplied...

Ideally the exception would be raised at the time the bad value was supplied.  But this would mean hooking into the etag attribute being set, or values being added to the cookies dict, which sounds like a drag.  So... The precedent set by `self.finished` handling in `write` is to raise an exception.  It would be good to stick with that, I guess.  So raising one of two new exception types sounds right to me.  Make sure to include enough information so that the developer can track down which header they screwed up.

Also I'm confused about the state of the unit tests for this.  Do the tests from the first patch still apply?  Or do some new ones need to be written?

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|ivank commented|
|-|-|

It was discovered[1] that `CR` could also be used to split headers in Chrome and IE, so the protection implemented here is not good enough.  Which also means that exarkun's suggestion above is even more reasonable, even though it's more work to implement.

Ignore the first patch; this needs new tests.

[1] http://bugs.php.net/bug.php?id=60227

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_11"></a>|@glyph set status to `closed`|
|-|-|

This was resolved by [#9420](https://github.com/twisted/twisted/issues/9420) by the approach of sanitizing what the spec describes as "linear whitespace", therefore resolving as a duplicate.

I wish that Jean-Paul's solution — quoting — had been the right one, but it turns out that the real correct solution that preserves maximum interoperability and follows the spirit of the spec is just mangling away all the whitespace.

HTTP headers are not the right place for your octet-accurate data, kids; short ASCII values only.  Stay in school.
